src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -485,6 +478,10 @@ private static Func<Task<TextAndVersion>> CreateTextAndVersionFunc(TextLoader te
                 {
                     oldSnapshot = entry.GetSnapshot();
                     newSnapshot = newEntry?.GetSnapshot() ?? oldSnapshot;
+
+                    // We're about to mutate, so assert that we're on the dispatcher thread.
+                    _dispatcher.AssertDispatcherThread();
```

Comment:
The dispatcher is still a requirement until it is removed. Essentially, it's an all or nothing situation. If nothing runs on the dispatcher, mutations are pushed directly onto the project snapshot manager queue. If everything runs on the dispatcher, mutations are pushed onto the dispatcher queue and are pushed onto the project snapshot manager queue later. However, if some mutations go directly to the project snapshot manager and some go to the dispatcher, mutations are likely to get processed in the wrong order because those on the dispatcher will be handled later.

Reply:
The addition of the lock allowed us to retrieve snapshot information without having to use the dispatcher, which was the win there. It looks like this just makes sure mutation happens on the dispatcher thread. That's not technically required from the perspective of the snapshot manager, but is logically required by dependents of snapshot changes. That's the part that will need to be cleaned up before we can completely remove

Reply:
> That's not technically required from the perspective of the snapshot manager.

It might not be required from the project snapshot manager's perspective, but it's required by the current system. Everything that calls mutation methods of the project snapshot manager does so using the dispatcher. By placing assertions here, I found a race condition where one component (the FallbackProjectManager) was updating the project snapshot manager directly without the dispatcher, before other components that had already queued updates on the dispatcher.

Reply:
> The addition of the lock allowed us to retrieve snapshot information without having to use the dispatcher, which was the win there.

Yes! And I really appreciate your work on this! The separation provided by the ReaderWriterLock and the refactoring you did made things so much clearer.